Exception when client requests 'attestation' events

Attempting to fetch attestation events from Teku stops after a while (usually 20-30 seconds).  The connection closes and an error is reported in the teku logs:

```
09:27:34.517 ERROR - Failed to process event queue for client 1759904277
java.util.concurrent.CompletionException: java.lang.IllegalStateException: AsyncContext completed and/or Request lifecycle recycled
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:331) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:346) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:632) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:141) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:494) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:28) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:138) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$25(SafeFuture.java:338) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2152) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:141) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:494) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:28) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:138) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateTo(SafeFuture.java:235) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:120) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: java.lang.IllegalStateException: AsyncContext completed and/or Request lifecycle recycled
        at org.eclipse.jetty.server.AsyncContextState.state(AsyncContextState.java:52) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
        at org.eclipse.jetty.server.AsyncContextState.getResponse(AsyncContextState.java:118) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
        at io.javalin.http.sse.Emitter.emit(Emitter.kt:34) ~[javalin-3.12.0.jar:?]
        at io.javalin.http.sse.SseClient.sendEvent(SseClient.java:29) ~[javalin-3.12.0.jar:?]
        at io.javalin.http.sse.SseClient.sendEvent(SseClient.java:25) ~[javalin-3.12.0.jar:?]
        at tech.pegasys.teku.beaconrestapi.handlers.v1.events.EventSubscriber.lambda$processEventQueue$0(EventSubscriber.java:82) ~[teku-data-beaconrestapi-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.fromRunnable(SafeFuture.java:150) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$0(AsyncRunner.java:24) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:69) ~[teku-infrastructure-async-21.1.1.jar:21.1.1]
        ... 4 more
```

### Steps to Reproduce (Bug)
1. run `curl -H "Accept:text/event-stream" 'http://localhost:5051/eth/v1/events?topics=attestation'`
2. wait a bit

**Expected behavior:** [What you expect to happen]
Stream continues until stopped.

**Actual behavior:** [What actually happens]
Stream is closed by Teku.

**Frequency:** [What percentage of the time does it occur?]

100%

### Versions (Add all that apply)

* Software version: [`teku --version`]

`teku/v21.1.1/linux-x86_64/-privatebuild-openjdk64bitservervm-java-14`


* Java version: [`java -version`]
```
openjdk 14.0.2 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+12-Ubuntu-120.04)OpenJDK 64-Bit Server VM (build 14.0.2+12-Ubuntu-120.04, mixed mode, sharing)
```
* OS Name & Version: [`cat /etc/*release`]
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=20.04DISTRIB_CODENAME=focal
DISTRIB_DESCRIPTION="Ubuntu 20.04.1 LTS"NAME="Ubuntu"
VERSION="20.04.1 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.1 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```
* Kernel Version: [`uname -a`]
```
Linux jack-d01 5.4.0-54-generic #60-Ubuntu SMP Fri Nov 6 10:37:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```

### Additional Information

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,3 +35,5 @@ For information on changes in released versions of Teku, see the [releases page]
 - Fixed `NoSuchElementException` that occurred during syncing.
 - Avoid marking the node as in sync incorrectly if an error occurs while syncing. Now selects a new target chain and continues syncing.
 - Reject validator related REST API requests when syncing, even if the head block is close to the current slot.  `head` and `chain_reorg` events do not fire while sync is active so the validator client is unable to detect when it should invalidate duties.
+- Increase the default limit for the queue for delivering events to REST API subscribers. When subscribing to attestations, they are often received in bursts which would exceed the previous limit even when the client was keeping up.
+  The default limit is now 250 and can now be configured with `--Xrest-api-max-pending-events`.

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/BeaconRestApi.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/BeaconRestApi.java
@@ -94,7 +94,6 @@ public class BeaconRestApi {
   private final Javalin app;
   private final JsonProvider jsonProvider = new JsonProvider();
   private static final Logger LOG = LogManager.getLogger();
-  public static final String FILE_NOT_FOUND_HTML = "404.html";
 
   private void initialize(
       final DataProvider dataProvider,
@@ -117,7 +116,7 @@ public class BeaconRestApi {
     addExceptionHandlers();
     // standard api endpoint inclusion
     addV1BeaconHandlers(dataProvider);
-    addEventHandler(dataProvider, eventChannels, asyncRunner);
+    addEventHandler(dataProvider, eventChannels, asyncRunner, configuration);
     addV1NodeHandlers(dataProvider);
     addV1ValidatorHandlers(dataProvider);
     addV1ConfigHandlers(dataProvider, configuration.getEth1DepositContractAddress());
@@ -335,8 +334,16 @@ public class BeaconRestApi {
   private void addEventHandler(
       final DataProvider dataProvider,
       final EventChannels eventChannels,
-      final AsyncRunner asyncRunner) {
-    app.get(GetEvents.ROUTE, new GetEvents(dataProvider, jsonProvider, eventChannels, asyncRunner));
+      final AsyncRunner asyncRunner,
+      final BeaconRestApiConfig configuration) {
+    app.get(
+        GetEvents.ROUTE,
+        new GetEvents(
+            dataProvider,
+            jsonProvider,
+            eventChannels,
+            asyncRunner,
+            configuration.getMaxPendingEvents()));
   }
 
   public void stop() {

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/BeaconRestApiConfig.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/BeaconRestApiConfig.java
@@ -28,6 +28,7 @@ public class BeaconRestApiConfig {
   private final List<String> restApiHostAllowlist;
   private final List<String> restApiCorsAllowedOrigins;
   private final Optional<Eth1Address> eth1DepositContractAddress;
+  private final int maxPendingEvents;
 
   private BeaconRestApiConfig(
       final int restApiPort,
@@ -36,7 +37,8 @@ public class BeaconRestApiConfig {
       final String restApiInterface,
       final List<String> restApiHostAllowlist,
       final List<String> restApiCorsAllowedOrigins,
-      final Optional<Eth1Address> eth1DepositContractAddress) {
+      final Optional<Eth1Address> eth1DepositContractAddress,
+      final int maxPendingEvents) {
     this.restApiPort = restApiPort;
     this.restApiDocsEnabled = restApiDocsEnabled;
     this.restApiEnabled = restApiEnabled;
@@ -44,6 +46,7 @@ public class BeaconRestApiConfig {
     this.restApiHostAllowlist = restApiHostAllowlist;
     this.restApiCorsAllowedOrigins = restApiCorsAllowedOrigins;
     this.eth1DepositContractAddress = eth1DepositContractAddress;
+    this.maxPendingEvents = maxPendingEvents;
   }
 
   public int getRestApiPort() {
@@ -74,6 +77,10 @@ public class BeaconRestApiConfig {
     return eth1DepositContractAddress;
   }
 
+  public int getMaxPendingEvents() {
+    return maxPendingEvents;
+  }
+
   public static BeaconRestApiConfigBuilder builder() {
     return new BeaconRestApiConfigBuilder();
   }
@@ -87,6 +94,7 @@ public class BeaconRestApiConfig {
     private List<String> restApiHostAllowlist;
     private List<String> restApiCorsAllowedOrigins;
     private Optional<Eth1Address> eth1DepositContractAddress = Optional.empty();
+    private int maxPendingEvents;
 
     private BeaconRestApiConfigBuilder() {}
 
@@ -136,6 +144,11 @@ public class BeaconRestApiConfig {
       return this;
     }
 
+    public BeaconRestApiConfigBuilder maxPendingEvents(final int maxEventQueueSize) {
+      this.maxPendingEvents = maxEventQueueSize;
+      return this;
+    }
+
     public BeaconRestApiConfig build() {
       return new BeaconRestApiConfig(
           restApiPort,
@@ -144,7 +157,8 @@ public class BeaconRestApiConfig {
           restApiInterface,
           restApiHostAllowlist,
           restApiCorsAllowedOrigins,
-          eth1DepositContractAddress);
+          eth1DepositContractAddress,
+          maxPendingEvents);
     }
   }
 }

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java
@@ -25,11 +25,12 @@ import tech.pegasys.teku.infrastructure.async.AsyncRunner;
 
 public class EventSubscriber {
   private static final Logger LOG = LogManager.getLogger();
-  static final int MAX_PENDING_EVENTS = 10;
+  private final AtomicBoolean stopped = new AtomicBoolean(false);
   private final List<EventType> eventTypes;
   private final SseClient sseClient;
   private final Queue<QueuedEvent> queuedEvents;
   private final Runnable closeCallback;
+  private final int maxPendingEvents;
   private final AtomicBoolean processingQueue;
   final AsyncRunner asyncRunner;
 
@@ -37,10 +38,12 @@ public class EventSubscriber {
       final List<String> eventTypes,
       final SseClient sseClient,
       final Runnable closeCallback,
-      final AsyncRunner asyncRunner) {
+      final AsyncRunner asyncRunner,
+      final int maxPendingEvents) {
     this.eventTypes = EventType.getTopics(eventTypes);
     this.sseClient = sseClient;
     this.closeCallback = closeCallback;
+    this.maxPendingEvents = maxPendingEvents;
     this.queuedEvents = new ConcurrentLinkedQueue<>();
     this.processingQueue = new AtomicBoolean(false);
     this.asyncRunner = asyncRunner;
@@ -51,11 +54,12 @@ public class EventSubscriber {
     if (!eventTypes.contains(eventType)) {
       return;
     }
-    if (queuedEvents.size() < MAX_PENDING_EVENTS) {
+    if (queuedEvents.size() < maxPendingEvents) {
       queuedEvents.add(QueuedEvent.of(eventType, message));
       processEventQueue();
     } else {
-      LOG.trace("Closing client connection due to exceeding the pending message limit");
+      LOG.debug("Closing event connection due to exceeding the pending message limit");
+      stopped.set(true);
       sseClient.ctx.req.getAsyncContext().complete();
       closeCallback.run();
     }
@@ -66,7 +70,7 @@ public class EventSubscriber {
   }
 
   private void processEventQueue() {
-    if (!processingQueue.compareAndSet(false, true)) {
+    if (!stopped.get() && !processingQueue.compareAndSet(false, true)) {
       // any queue processing in progress will clear the queue, no need to run another instance
       return;
     }
@@ -78,7 +82,7 @@ public class EventSubscriber {
                   queuedEvents.size(),
                   sseClient.hashCode());
               QueuedEvent event = queuedEvents.poll();
-              while (event != null) {
+              while (event != null && !stopped.get()) {
                 sseClient.sendEvent(event.getEventType().name(), event.getMessageData());
                 event = queuedEvents.poll();
               }

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriptionManager.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriptionManager.java
@@ -56,6 +56,7 @@ public class EventSubscriptionManager implements ChainHeadChannel, FinalizedChec
   private final JsonProvider jsonProvider;
   private final ChainDataProvider provider;
   private final AsyncRunner asyncRunner;
+  private final int maxPendingEvents;
   // collection of subscribers
   private final Collection<EventSubscriber> eventSubscribers;
 
@@ -66,10 +67,12 @@ public class EventSubscriptionManager implements ChainHeadChannel, FinalizedChec
       final SyncDataProvider syncDataProvider,
       final ConfigProvider configProvider,
       final AsyncRunner asyncRunner,
-      final EventChannels eventChannels) {
+      final EventChannels eventChannels,
+      final int maxPendingEvents) {
     this.provider = chainDataProvider;
     this.jsonProvider = jsonProvider;
     this.asyncRunner = asyncRunner;
+    this.maxPendingEvents = maxPendingEvents;
     this.eventSubscribers = new ConcurrentLinkedQueue<>();
     this.configProvider = configProvider;
     eventChannels.subscribe(ChainHeadChannel.class, this);
@@ -92,7 +95,8 @@ public class EventSubscriptionManager implements ChainHeadChannel, FinalizedChec
               eventSubscribers.removeIf(sub -> sub.getSseClient().equals(sseClient));
               LOG.trace("disconnected " + sseClient.hashCode());
             },
-            asyncRunner);
+            asyncRunner,
+            maxPendingEvents);
     eventSubscribers.add(subscriber);
   }
 

--- a/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/GetEvents.java
+++ b/data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/GetEvents.java
@@ -54,7 +54,8 @@ public class GetEvents implements Handler {
       final DataProvider dataProvider,
       final JsonProvider jsonProvider,
       final EventChannels eventChannels,
-      final AsyncRunner asyncRunner) {
+      final AsyncRunner asyncRunner,
+      final int maxPendingEvents) {
     this(
         dataProvider.getNodeDataProvider(),
         dataProvider.getChainDataProvider(),
@@ -62,7 +63,8 @@ public class GetEvents implements Handler {
         dataProvider.getSyncDataProvider(),
         dataProvider.getConfigProvider(),
         eventChannels,
-        asyncRunner);
+        asyncRunner,
+        maxPendingEvents);
   }
 
   GetEvents(
@@ -72,7 +74,8 @@ public class GetEvents implements Handler {
       final SyncDataProvider syncDataProvider,
       final ConfigProvider configProvider,
       final EventChannels eventChannels,
-      final AsyncRunner asyncRunner) {
+      final AsyncRunner asyncRunner,
+      final int maxPendingEvents) {
     this.jsonProvider = jsonProvider;
     eventSubscriptionManager =
         new EventSubscriptionManager(
@@ -82,7 +85,8 @@ public class GetEvents implements Handler {
             syncDataProvider,
             configProvider,
             asyncRunner,
-            eventChannels);
+            eventChannels,
+            maxPendingEvents);
   }
 
   @OpenApi(

--- a/data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriberTest.java
+++ b/data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriberTest.java
@@ -20,7 +20,6 @@ import static org.mockito.Mockito.never;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
-import static tech.pegasys.teku.beaconrestapi.handlers.v1.events.EventSubscriber.MAX_PENDING_EVENTS;
 
 import io.javalin.http.Context;
 import io.javalin.http.sse.SseClient;
@@ -42,6 +41,7 @@ import tech.pegasys.teku.api.response.v1.EventType;
 import tech.pegasys.teku.infrastructure.async.StubAsyncRunner;
 
 public class EventSubscriberTest {
+  private static final int MAX_PENDING_EVENTS = 10;
   private final AsyncContext asyncContext = mock(AsyncContext.class);
   private final HttpServletRequest req = mock(HttpServletRequest.class);
   private final HttpServletResponse res = mock(HttpServletResponse.class);
@@ -67,14 +67,16 @@ public class EventSubscriberTest {
   @Test
   void shouldGetSseClient() {
     EventSubscriber eventSubscriber =
-        new EventSubscriber(List.of("head"), sseClient, onCloseCallback, asyncRunner);
+        new EventSubscriber(
+            List.of("head"), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
     assertThat(eventSubscriber.getSseClient()).isEqualTo(sseClient);
   }
 
   @Test
   void shouldDisconnectAfterTooManyRequestsAreLogged() {
     EventSubscriber eventSubscriber =
-        new EventSubscriber(List.of("head"), sseClient, onCloseCallback, asyncRunner);
+        new EventSubscriber(
+            List.of("head"), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
 
     for (int i = 0; i < MAX_PENDING_EVENTS + 1; i++) {
       verify(onCloseCallback, never()).run();
@@ -84,13 +86,30 @@ public class EventSubscriberTest {
   }
 
   @Test
+  void shouldStopSendingEventsWhenQueueOverflows() throws Exception {
+    EventSubscriber eventSubscriber =
+        new EventSubscriber(
+            List.of("head"), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
+
+    for (int i = 0; i < MAX_PENDING_EVENTS + 1; i++) {
+      verify(onCloseCallback, never()).run();
+      eventSubscriber.onEvent(EventType.head, "test");
+    }
+    verify(onCloseCallback).run();
+    verify(asyncContext).complete();
+    asyncRunner.executeQueuedActions();
+    verify(outputStream, never()).print(anyString());
+  }
+
+  @Test
   void shouldSubscribeToMultipleEventsSuccessfully() throws IOException {
     EventSubscriber eventSubscriber =
         new EventSubscriber(
             allEventTypes.stream().map(EventType::name).collect(Collectors.toList()),
             sseClient,
             onCloseCallback,
-            asyncRunner);
+            asyncRunner,
+            MAX_PENDING_EVENTS);
     allEventTypes.forEach(eventType -> eventSubscriber.onEvent(eventType, "test"));
     assertThat(asyncRunner.countDelayedActions()).isEqualTo(1);
     asyncRunner.executeQueuedActions();
@@ -100,7 +119,8 @@ public class EventSubscriberTest {
   @Test
   void shouldNotDisconnectIfQueueProcessingCatchesUp() throws IOException {
     EventSubscriber eventSubscriber =
-        new EventSubscriber(List.of("head"), sseClient, onCloseCallback, asyncRunner);
+        new EventSubscriber(
+            List.of("head"), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
 
     for (int i = 0; i < MAX_PENDING_EVENTS; i++) {
       eventSubscriber.onEvent(EventType.head, "test");
@@ -119,7 +139,8 @@ public class EventSubscriberTest {
   @EnumSource(EventType.class)
   void shouldNotSendEventsIfNotSubscribed(final EventType eventType) {
     EventSubscriber subscriber =
-        new EventSubscriber(List.of(eventType.name()), sseClient, onCloseCallback, asyncRunner);
+        new EventSubscriber(
+            List.of(eventType.name()), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
     allEventTypes.stream()
         .filter(val -> val.compareTo(eventType) != 0)
         .forEach(value -> subscriber.onEvent(value, "test"));
@@ -130,7 +151,8 @@ public class EventSubscriberTest {
   @EnumSource(EventType.class)
   void shouldSendEventsIfSubscribed(final EventType eventType) throws IOException {
     EventSubscriber subscriber =
-        new EventSubscriber(List.of(eventType.name()), sseClient, onCloseCallback, asyncRunner);
+        new EventSubscriber(
+            List.of(eventType.name()), sseClient, onCloseCallback, asyncRunner, MAX_PENDING_EVENTS);
 
     subscriber.onEvent(eventType, "test");
 

--- a/data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriptionManagerTest.java
+++ b/data/beaconrestapi/src/test/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriptionManagerTest.java
@@ -128,7 +128,8 @@ public class EventSubscriptionManagerTest {
             syncDataProvider,
             configProvider,
             asyncRunner,
-            channels);
+            channels,
+            10);
     client1 = new SseClient(ctx);
   }
 

--- a/teku/src/main/java/tech/pegasys/teku/cli/options/BeaconRestApiOptions.java
+++ b/teku/src/main/java/tech/pegasys/teku/cli/options/BeaconRestApiOptions.java
@@ -22,6 +22,7 @@ import tech.pegasys.teku.config.TekuConfiguration;
 public class BeaconRestApiOptions {
 
   public static final int DEFAULT_REST_API_PORT = 5051;
+  public static final int DEFAULT_MAX_EVENT_QUEUE_SIZE = 250;
 
   @Option(
       names = {"--rest-api-port"},
@@ -69,29 +70,11 @@ public class BeaconRestApiOptions {
       arity = "0..*")
   private final List<String> restApiCorsAllowedOrigins = new ArrayList<>();
 
-  public int getRestApiPort() {
-    return restApiPort;
-  }
-
-  public boolean isRestApiDocsEnabled() {
-    return restApiDocsEnabled;
-  }
-
-  public boolean isRestApiEnabled() {
-    return restApiEnabled;
-  }
-
-  public String getRestApiInterface() {
-    return restApiInterface;
-  }
-
-  public List<String> getRestApiHostAllowlist() {
-    return restApiHostAllowlist;
-  }
-
-  public List<String> getRestApiCorsAllowedOrigins() {
-    return restApiCorsAllowedOrigins;
-  }
+  @Option(
+      names = {"--Xrest-api-max-pending-events"},
+      paramLabel = "<INTEGER>",
+      hidden = true)
+  private int maxPendingEvents = DEFAULT_MAX_EVENT_QUEUE_SIZE;
 
   public void configure(final TekuConfiguration.Builder builder) {
     builder.restApi(
@@ -102,6 +85,7 @@ public class BeaconRestApiOptions {
                 .restApiPort(restApiPort)
                 .restApiInterface(restApiInterface)
                 .restApiHostAllowlist(restApiHostAllowlist)
-                .restApiCorsAllowedOrigins(restApiCorsAllowedOrigins));
+                .restApiCorsAllowedOrigins(restApiCorsAllowedOrigins)
+                .maxPendingEvents(maxPendingEvents));
   }
 }

--- a/teku/src/test/java/tech/pegasys/teku/cli/BeaconNodeCommandTest.java
+++ b/teku/src/test/java/tech/pegasys/teku/cli/BeaconNodeCommandTest.java
@@ -45,6 +45,7 @@ import org.apache.commons.lang3.StringUtils;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
+import tech.pegasys.teku.cli.options.BeaconRestApiOptions;
 import tech.pegasys.teku.config.TekuConfiguration;
 import tech.pegasys.teku.datastructures.eth1.Eth1Address;
 import tech.pegasys.teku.infrastructure.version.VersionProvider;
@@ -437,7 +438,8 @@ public class BeaconNodeCommandTest extends AbstractBeaconNodeCommandTest {
                     .restApiInterface("127.0.0.1")
                     .restApiHostAllowlist(List.of("127.0.0.1", "localhost"))
                     .restApiCorsAllowedOrigins(new ArrayList<>())
-                    .eth1DepositContractAddress(Optional.of(address)))
+                    .eth1DepositContractAddress(Optional.of(address))
+                    .maxPendingEvents(BeaconRestApiOptions.DEFAULT_MAX_EVENT_QUEUE_SIZE))
         .validator(
             b ->
                 b.validatorExternalSignerTimeout(Duration.ofSeconds(5))

